README
c-utf8 - UTF-8 Handling in Standard ISO-C11

ABOUT:
        The c-utf8 project implements utility functions around handling UTF-8
        in Standard ISO-C11.

        For API documentation, see the c-utf8.h header file, as well as the
        docbook comments for each function.

DETAILS:
        https://c-util.github.io/c-utf8

BUG REPORTS:
        https://github.com/c-util/c-utf8/issues

GIT:
        git@github.com:c-util/c-utf8.git
        https://github.com/c-util/c-utf8.git

GITWEB:
        https://github.com/c-util/c-utf8

LICENSE:
        Apache Software License 2.0
        Lesser General Public License 2.1+
        See AUTHORS for details.

REQUIREMENTS:
        The requirements for c-utf8 are:

            libc (e.g., glibc >= 2.16)

        At build-time, the following software is required:

            meson >= 0.41
            pkg-config >= 0.29

INSTALL:
        The meson build-system is used for this project. Contact upstream
        documentation for detailed help. In most situations the following
        commands are sufficient to build and install from source:

            $ mkdir build
            $ cd build
            $ meson setup ..
            $ ninja
            $ meson test
            # ninja install

        No custom configuration options are available.


NEWS.md
# c-utf8 - UTF-8 Handling in Standard ISO-C11

## CHANGES WITH 1.1.0:

        * Align with strict-aliasing rules to fix miscompilations whith flags
          like `-fstrict-aliasing`.

        Contributions from: David Rheinsberg, Jan Engelhardt, Tom Gundersen

        - Dußlingen, 2023-12-12

## CHANGES WITH 1.0.0:

        * Initial release of c-utf8.

        Contributions from: David Rheinsberg, Evgeny Vereshchagin, Tom Gundersen

        - Brno, 2022-06-22


src/test-basic.c
/*
 * Tests for UTF-8 validation helper
 */

#undef NDEBUG
#include <assert.h>
#include <c-stdaux.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include "c-utf8.h"

static void test_ascii(void) {
        char str[0x100];
        const char *p = str;
        size_t len = sizeof(str);

        for (size_t i = 0; i < sizeof(str); ++i)
                str[i] = i;

        c_utf8_verify_ascii(&p, &len);
        c_assert(*p == 0x00);
        c_assert(p == str);
        c_assert(len == sizeof(str));

        ++p;
        --len;

        c_utf8_verify_ascii(&p, &len);
        c_assert((unsigned char)*p == 0x80);
        c_assert(p == str + 0x7F + 1);
        c_assert(len == sizeof(str) - 0x7F - 1);

}

static void test_utf8(void) {
        /* verify a mix of greek, czech and chinese */
        {
                char str[] = "Η Ελλάδα ή Ελλάς, επίσημα γνωστή ως Ελληνική "
                             "Δημοκρατία, είναι χώρα της νοτιοανατολικής "
                             "Ευρώπης στο νοτιότερο άκρο της Βαλκανικής "
                             "χερσονήσου. Συνορεύει στα βορειοδυτικά με την "
                             "Αλβανία, στα βόρεια με την πρώην Γιουγκοσλαβική "
                             "Δημοκρατία της Μακεδονίας και τη Βουλγαρία και "
                             "στα βορειοανατολικά με την Τουρκία. Česko, "
                             "úředním názvem Česká republika, je stát ve "
                             "střední Evropě. Jako formálně svrchovaný národní "
                             "stát vznikla tehdejší Česká socialistická "
                             "republika 1. ledna 1969 v rámci federalizace "
                             "Československa. Od 6. března 1990 nese tento "
                             "stát název Česká republika. 中華民國十年，"
                             "中國共產黨立於上海。初附於中國國民黨，"
                             "黨人得以兼國民黨，共理中華民國廣州軍政府，"
                             "同謀北伐。其後國民黨人以共產黨人以公務營黨務，"
                             "既下南京，蔣中正令捕殺共產黨人。遂奔江西。"
                             "十六年起義於南昌，中國工農紅軍是立";
                const char *p = str;
                size_t len = sizeof(str);

                c_utf8_verify(&p, &len);
                c_assert(len == 1);
                c_assert(p == str + strlen(str));
                c_assert(*p == 0x00);
        }

        /* verify every 1-byte character */
        for (uint32_t i = 0; i <= 0xFF; i++) {
                char str[] = { i };
                size_t len = sizeof(str);
                const char *p = str;

                c_utf8_verify(&p, &len);
                if (str[0] == 0 || (str[0] & 0b10000000) != 0)
                        c_assert(p == str && len == sizeof(str));
                else
                        c_assert(p == str + sizeof(str) && len == 0);
        }

        /* verify every 2-byte character */
        for (uint32_t i = 0; i <= 0xFFFF; i++) {
                char str[] = { (i >> 8), i & 0xFF };
                size_t len = sizeof(str);
                const char *p = str;
                uint32_t n;

                if ((str[0] & 0x80) == 0)
                        /* ignore leading 1-byte characters */
                        continue;

                n = (str[0] & 0b00011111) << 6 | (str[1] & 0b00111111);

                c_utf8_verify(&p, &len);

                if (((str[0] & 0b11100000) != 0b11000000) ||
                    ((str[1] & 0b11000000) != 0b10000000) ||
                    (n < 0x80)) /* overlong */
                        c_assert(p == str && len == sizeof(str));
                else
                        c_assert(p == str + sizeof(str) && len == 0);
        }

        /* verify every 0xF'th 3-byte character */
        for (uint32_t i = 0; i <= 0xFFFFFF; i += 0xF) {
                char str[] = { (i >> 16), (i >> 8) & 0xFF, i & 0xFF };
                size_t len = sizeof(str);
                const char *p = str;
                uint32_t n;

                if (((str[0] & 0b10000000) == 0) ||
                    ((str[0] & 0b11100000) == 0b11000000))
                        /* ignore leading 1,2-byte characters */
                        continue;

                n = (str[0] & 0b00001111) << 12 | (str[1] & 0b00111111) << 6 | (str[2] & 0b00111111);

                c_utf8_verify(&p, &len);

                if (((str[0] & 0b11110000) != 0b11100000) ||
                    ((str[1] & 0b11000000) != 0b10000000) ||
                    ((str[2] & 0b11000000) != 0b10000000) ||
                    (n < 0x800) || /* overlong */
                    (n >= 0xD800 && n <= 0xDFFF)) /* surrogates */
                        c_assert(p == str && len == sizeof(str));
                else
                        c_assert(p == str + sizeof(str) && len == 0);
        }

        /* verify every 0xFFF'th 4-byte character */
        for (uint64_t i = 0; i <= 0xFFFFFFFF; i += 0xFFF) {
                char str[] = { (i >> 24), (i >> 16) & 0xFF, (i >> 8) & 0xFF, i & 0xFF };
                size_t len = sizeof(str);
                const char *p = str;
                uint32_t n;

                if (((str[0] & 0b10000000) == 0) ||
                    ((str[0] & 0b11100000) == 0b11000000) ||
                    ((str[0] & 0b11110000) == 0b11100000))
                        /* ignore leading 1,2,3-byte characters */
                        continue;

                n = (str[0] & 0b00000111) << 18 | (str[1] & 0b00111111) << 12 | (str[2] & 0b00111111) << 6 | (str[3] & 0b00111111);

                c_utf8_verify(&p, &len);

                if (((str[0] & 0b11111000) != 0b11110000) ||
                    ((str[1] & 0b11000000) != 0b10000000) ||
                    ((str[2] & 0b11000000) != 0b10000000) ||
                    ((str[3] & 0b11000000) != 0b10000000) ||
                    (n < 0x10000) || /* overlong */
                    (n > 0x10FFFF)) /* out of unicode range */
                        c_assert(p == str && len == sizeof(str));
                else
                        c_assert(p == str + sizeof(str) && len == 0);
        }
}

int main(int argc, char **argv) {
        test_ascii();
        test_utf8();
        return 0;
}


src/bench-ascii.c
/*
 * ASCII validation benchmark
 */

#undef NDEBUG
#include <assert.h>
#include <c-stdaux.h>
#include <inttypes.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include "c-utf8.h"
#include "trivial-utf8.h"

static uint64_t test_get_time(void) {
        struct timespec ts;
        int r;

        r = clock_gettime(CLOCK_THREAD_CPUTIME_ID, &ts);
        c_assert(r >= 0);

        return ts.tv_sec * UINT64_C(1000000000) + ts.tv_nsec;
}

static void test_trivial_utf8(const char *string, size_t n_string) {
        uint64_t ts;
        const char *res;

        ts = test_get_time();
        res = trivial_utf8_is_valid(string);
        fprintf(stderr, "UTF-8 verify ASCII string (trivial): %"PRIu64" MB/s\n", n_string * 1000 * 1000 * 1000 / 1024 / 1024 / (test_get_time() - ts));
        c_assert(res == string);
}

static void test_utf8(const char *string, size_t n_string) {
        uint64_t ts;
        size_t len = n_string;

        ts = test_get_time();
        c_utf8_verify(&string, &len);
        fprintf(stderr, "UTF-8 verify ASCII string: %"PRIu64" MB/s\n", n_string * 1000 * 1000 * 1000 / 1024 / 1024 / (test_get_time() - ts));
        c_assert(len == 1);
}

static void test_strlen(const char *string, size_t n_string) {
        uint64_t ts;
        size_t len;

        ts = test_get_time();
        len = strnlen(string, n_string);
        fprintf(stderr, "strlen() of ASCII string: %"PRIu64" MB/s\n", n_string * 1000 * 1000 * 1000 / 1024 / 1024 / (test_get_time() - ts));
        c_assert(len == n_string - 1);
}

#define TEST_STRING_SIZE (500ULL * 1024ULL * 1024ULL)

int main(int argc, char **argv) {
        char *string;

        string = malloc(TEST_STRING_SIZE);
        c_assert(string);

        for (size_t i = 0; i < TEST_STRING_SIZE - 1; ++i)
                string[i] = (i % 127) + 1;

        string[TEST_STRING_SIZE - 1] = '\0';

        test_trivial_utf8(string, TEST_STRING_SIZE);
        test_utf8(string, TEST_STRING_SIZE);
        test_strlen(string, TEST_STRING_SIZE);

        return 0;
}


src/trivial-utf8.c
/* SPDX-License-Identifier: LGPL-2.1+ */
/***
  This file is part of systemd.

  Copyright 2008-2011 Kay Sievers
  Copyright 2012 Lennart Poettering

  systemd is free software; you can redistribute it and/or modify it
  under the terms of the GNU Lesser General Public License as published by
  the Free Software Foundation; either version 2.1 of the License, or
  (at your option) any later version.

  systemd is distributed in the hope that it will be useful, but
  WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
  Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public License
  along with systemd; If not, see <http://www.gnu.org/licenses/>.
***/

/* Parts of this file are based on the GLIB utf8 validation functions. The
 * original license text follows. */

/* gutf8.c - Operations on UTF-8 strings.
 *
 * Copyright (C) 1999 Tom Tromey
 * Copyright (C) 2000 Red Hat, Inc.
 *
 * This library is free software; you can redistribute it and/or
 * modify it under the terms of the GNU Library General Public
 * License as published by the Free Software Foundation; either
 * version 2 of the License, or (at your option) any later version.
 *
 * This library is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
 * Library General Public License for more details.
 *
 * You should have received a copy of the GNU Library General Public
 * License along with this library; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301  USA
 */

#include <c-stdaux.h>
#include <errno.h>
#include <stdbool.h>
#include <stddef.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <uchar.h>
#include "trivial-utf8.h"

/* count of characters used to encode one unicode char */
static int utf8_encoded_expected_len(const char *str) {
        unsigned char c;

        c = (unsigned char) str[0];
        if (c < 0x80)
                return 1;
        if ((c & 0xe0) == 0xc0)
                return 2;
        if ((c & 0xf0) == 0xe0)
                return 3;
        if ((c & 0xf8) == 0xf0)
                return 4;
        if ((c & 0xfc) == 0xf8)
                return 5;
        if ((c & 0xfe) == 0xfc)
                return 6;

        return 0;
}

/* decode one unicode char */
static int utf8_encoded_to_unichar(const char *str, char32_t *ret_unichar) {
        char32_t unichar;
        int len, i;

        len = utf8_encoded_expected_len(str);

        switch (len) {
        case 1:
                *ret_unichar = (char32_t)str[0];
                return 0;
        case 2:
                unichar = str[0] & 0x1f;
                break;
        case 3:
                unichar = (char32_t)str[0] & 0x0f;
                break;
        case 4:
                unichar = (char32_t)str[0] & 0x07;
                break;
        case 5:
                unichar = (char32_t)str[0] & 0x03;
                break;
        case 6:
                unichar = (char32_t)str[0] & 0x01;
                break;
        default:
                return -EINVAL;
        }

        for (i = 1; i < len; i++) {
                if (((char32_t)str[i] & 0xc0) != 0x80)
                        return -EINVAL;
                unichar <<= 6;
                unichar |= (char32_t)str[i] & 0x3f;
        }

        *ret_unichar = unichar;

        return 0;
}

/* expected size used to encode one unicode char */
static int utf8_unichar_to_encoded_len(char32_t unichar) {

        if (unichar < 0x80)
                return 1;
        if (unichar < 0x800)
                return 2;
        if (unichar < 0x10000)
                return 3;
        if (unichar < 0x200000)
                return 4;
        if (unichar < 0x4000000)
                return 5;

        return 6;
}

static bool unichar_is_valid(char32_t ch) {

        if (ch >= 0x110000) /* End of unicode space */
                return false;
        if ((ch & 0xFFFFF800) == 0xD800) /* Reserved area for UTF-16 */
                return false;

        return true;
}

/* validate one encoded unicode char and return its length */
static int utf8_encoded_valid_unichar(const char *str) {
        int len, i, r;
        char32_t unichar;

        len = utf8_encoded_expected_len(str);
        if (len == 0)
                return -EINVAL;

        /* ascii is valid */
        if (len == 1)
                return 1;

        /* check if expected encoded chars are available */
        for (i = 0; i < len; i++)
                if ((str[i] & 0x80) != 0x80)
                        return -EINVAL;

        r = utf8_encoded_to_unichar(str, &unichar);
        if (r < 0)
                return r;

        /* check if encoded length matches encoded value */
        if (utf8_unichar_to_encoded_len(unichar) != len)
                return -EINVAL;

        /* check if value has valid range */
        if (!unichar_is_valid(unichar))
                return -EINVAL;

        return len;
}

const char *trivial_utf8_is_valid(const char *str) {
        const uint8_t *p;

        for (p = (const uint8_t*) str; *p; ) {
                int len;

                len = utf8_encoded_valid_unichar((const char *)p);
                if (len < 0) {
                        fprintf(stderr, "failed after %zd characters\n", (void*)p - (void*)str);
                        return NULL;
                }

                p += len;
        }

        return str;
}

const char *trivial_ascii_is_valid(const char *str) {
        const char *p;

        for (p = str; *p; p++)
                if ((unsigned char) *p >= 128)
                        return NULL;

        return (char*) str;
}


src/bench-multibyte.c
/*
 * UTF-8 multibyte validation benchmark
 */

#undef NDEBUG
#include <assert.h>
#include <c-stdaux.h>
#include <inttypes.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <time.h>
#include "c-utf8.h"
#include "trivial-utf8.h"

static uint64_t test_get_time(void) {
        struct timespec ts;
        int r;

        r = clock_gettime(CLOCK_THREAD_CPUTIME_ID, &ts);
        c_assert(r >= 0);

        return ts.tv_sec * UINT64_C(1000000000) + ts.tv_nsec;
}

static void test_trivial_utf8(const char *string, size_t n_string, size_t n_bytes) {
        uint64_t ts;
        const char *res;

        ts = test_get_time();
        res = trivial_utf8_is_valid(string);
        fprintf(stderr, "UTF-8 verify string of %zu-byte characters (trivial): %"PRIu64" Mchar/s\n", n_bytes, n_string / n_bytes * 1000 * 1000 * 1000 / 1024 / 1024 / (test_get_time() - ts));
        c_assert(res == string);
}

static void test_utf8(const char *string, size_t n_string, size_t n_bytes) {
        uint64_t ts;
        size_t len = n_string;

        ts = test_get_time();
        c_utf8_verify(&string, &len);
        fprintf(stderr, "UTF-8 verify string of %zu-byte characters: %"PRIu64" Mchar/s\n", n_bytes, n_string / n_bytes * 1000 * 1000 * 1000 / 1024 / 1024 / (test_get_time() - ts));
        c_assert(len == 1);
}

#define TEST_STRING_SIZE (500ULL * 1024ULL * 1024ULL)

static void test_multibytes(size_t n_bytes) {
        char *string;

        string = calloc(1, TEST_STRING_SIZE);
        c_assert(string);

        for (size_t i = 0; i < (TEST_STRING_SIZE - 1) / n_bytes; ++i) {
                switch (n_bytes) {
                case 2: {
                        uint32_t c = (i % (0x800 - 0x80)) + 0x80;

                        string[i * n_bytes] = 0b11000000 | (c >> 6);
                        string[i * n_bytes + 1] = 0b10000000 | (c & 0b111111);

                        break;
                }
                case 3: {
                        uint32_t c = (i % (0x10000 - 0x800)) + 0x800;

                        if (c >= 0xD800 && c <= 0xDFFF)
                                c = 0x800;

                        string[i * n_bytes] = 0b11100000 | (c >> 12);
                        string[i * n_bytes + 1] = 0b10000000 | ((c >> 6) & 0b111111);
                        string[i * n_bytes + 2] = 0b10000000 | (c & 0b111111);

                        break;
                }
                case 4: {
                        uint32_t c = (i % (0x110000 - 0x10000)) + 0x10000;

                        string[i * n_bytes] = 0b11110000 | (c >> 18);
                        string[i * n_bytes + 1] = 0b10000000 | ((c >> 12) & 0b111111);
                        string[i * n_bytes + 2] = 0b10000000 | ((c >> 6) & 0b111111);
                        string[i * n_bytes + 3] = 0b10000000 | (c & 0b111111);

                        break;
                }
                default:
                        c_assert(0);
                }
        }

        test_trivial_utf8(string, ((TEST_STRING_SIZE - 1) / n_bytes) * n_bytes + 1, n_bytes);
        test_utf8(string, ((TEST_STRING_SIZE - 1) / n_bytes) * n_bytes + 1, n_bytes);

        free(string);
}

int main(int argc, char **argv) {
        test_multibytes(2);
        test_multibytes(3);
        test_multibytes(4);

        return 0;
}


src/c-utf8.c
/*
 * UTF-8 Implementation
 *
 * For highlevel documentation of the API see the header file and the docbook
 * comments. This implementation is inspired in part by Rust's std::str.
 *
 * So far only validation helpers are implemented, as those seem the most
 * critical.
 */

#include <c-stdaux.h>
#include <stddef.h>
#include <stdint.h>
#include "c-utf8.h"

/* The following constants are truncated on 32-bit machines */
#define C_UTF8_ASCII_MASK ((size_t)UINT64_C(0x8080808080808080))
#define C_UTF8_ASCII_SUB ((size_t)UINT64_C(0x0101010101010101))

static inline int c_utf8_word_is_ascii(size_t word) {
        /* True unless any byte is NULL or has the MSB set. */
        return ((((word - C_UTF8_ASCII_SUB) | word) & C_UTF8_ASCII_MASK) == 0);
}

/**
 * c_utf8_verify_ascii() - verify that a string is ASCII encoded
 * @strp:               pointer to string to verify
 * @lenp:               pointer to length of string
 *
 * Up to the first @lenp bytes of the string pointed to by @strp is
 * verified to be ASCII encoded, and @strp and @lenp are updated to
 * point to the first non-ASCII character or the first NULL of the
 * string, and the remaining number of bytes of the string,
 * respectively.
 *
 * If @lenp is NULL the string is scanned until the first invalid
 * byte, without any upper bound on its length.
 */
_c_public_ void c_utf8_verify_ascii(const char **strp, size_t *lenp) {
        const char *str = *strp;
        size_t len = lenp ? *lenp : (size_t)-1;

        while (len > 0 && c_load_8(str, 0) < 128) {
                if ((void*)c_align_to((unsigned long)str, sizeof(size_t)) == str) {
                        /*
                         * If the string is aligned to a word boundary, scan two
                         * words at a time for any NULL or non-ASCII characters.
                         *
                         * We do two words at a time to take advantage of the
                         * compiler being able to use SIMD instructions where
                         * available.
                         */
                        while (len >= 2 * sizeof(size_t)) {
                                if (!c_utf8_word_is_ascii(c_load(size_t, le, aligned, str, 0)) ||
                                    !c_utf8_word_is_ascii(c_load(size_t, le, aligned, str, sizeof(size_t))))
                                        break;

                                str += 2 * sizeof(size_t);
                                len -= 2 * sizeof(size_t);
                        }


                        /*
                         * Find the actual end of the ASCII-portion of the string.
                         */
                        while (len > 0 && c_load_8(str, 0) < 128) {
                                if (_c_unlikely_(c_load_8(str, 0) == 0x00))
                                        goto out;
                                ++str;
                                --len;
                        }
                } else {
                        /*
                         * The string was not aligned, scan one character at a time until
                         * it is.
                         */
                        if (_c_unlikely_(c_load_8(str, 0) == 0x00))
                                goto out;
                        ++str;
                        --len;
                }
        }

out:
        *strp = str;
        if (lenp)
                *lenp = len;
}

#define C_UTF8_CHAR_IS_TAIL(_x)         (((_x) & 0xC0) == 0x80)

/**
 * c_utf8_verify() - verify that a string is UTF-8 encoded
 * @strp:               pointer to string to verify
 * @lenp:               pointer to length of string, or NULL
 *
 * Up to the first @lenp bytes of the string pointed to by @strp is
 * verified to be UTF-8 encoded, and @strp and @lenp are updated to
 * point to the first non-UTF-8 character or the first NULL of the
 * string, and the remaining number of bytes of the string,
 * respectively.
 *
 * If @lenp is NULL the string is scanned until the first invalid
 * byte, without any upper bound on its length.
 */
_c_public_ void c_utf8_verify(const char **strp, size_t *lenp) {
        const char *str = *strp;
        size_t len = lenp ? *lenp : (size_t)-1;

        /* See Unicode 10.0.0, Chapter 3, Section D92 */

        while (len > 0) {
                switch (c_load_8(str, 0)) {
                case 0x00:
                        goto out;
                case 0x01 ... 0x7F:
                        /*
                         * Special-case and optimize the ASCII case.
                         */
                        c_utf8_verify_ascii((const char **)&str, &len);

                        break;
                case 0xC2 ... 0xDF:
                        if (_c_unlikely_(len < 2))
                                goto out;
                        if (_c_unlikely_(!C_UTF8_CHAR_IS_TAIL(c_load_8(str, 1))))
                                goto out;

                        str += 2;
                        len -= 2;

                        break;
                case 0xE0:
                        if (_c_unlikely_(len < 3))
                                goto out;
                        if (_c_unlikely_(c_load_8(str, 1) < 0xA0 || c_load_8(str, 1) > 0xBF))
                                goto out;
                        if (_c_unlikely_(!C_UTF8_CHAR_IS_TAIL(c_load_8(str, 2))))
                                goto out;

                        str += 3;
                        len -= 3;

                        break;
                case 0xE1 ... 0xEC:
                        if (_c_unlikely_(len < 3))
                                goto out;
                        if (_c_unlikely_(!C_UTF8_CHAR_IS_TAIL(c_load_8(str, 1))))
                                goto out;
                        if (_c_unlikely_(!C_UTF8_CHAR_IS_TAIL(c_load_8(str, 2))))
                                goto out;

                        str += 3;
                        len -= 3;

                        break;
                case 0xED:
                        if (_c_unlikely_(len < 3))
                                goto out;
                        if (_c_unlikely_(c_load_8(str, 1) < 0x80 || c_load_8(str, 1) > 0x9F))
                                goto out;
                        if (_c_unlikely_(!C_UTF8_CHAR_IS_TAIL(c_load_8(str, 2))))
                                goto out;

                        str += 3;
                        len -= 3;

                        break;
                case 0xEE ... 0xEF:
                        if (_c_unlikely_(len < 3))
                                goto out;
                        if (_c_unlikely_(!C_UTF8_CHAR_IS_TAIL(c_load_8(str, 1))))
                                goto out;
                        if (_c_unlikely_(!C_UTF8_CHAR_IS_TAIL(c_load_8(str, 2))))
                                goto out;

                        str += 3;
                        len -= 3;

                        break;
                case 0xF0:
                        if (_c_unlikely_(len < 4))
                                goto out;
                        if (_c_unlikely_(c_load_8(str, 1) < 0x90 || c_load_8(str, 1) > 0xBF))
                                goto out;
                        if (_c_unlikely_(!C_UTF8_CHAR_IS_TAIL(c_load_8(str, 2))))
                                goto out;
                        if (_c_unlikely_(!C_UTF8_CHAR_IS_TAIL(c_load_8(str, 3))))
                                goto out;

                        str += 4;
                        len -= 4;

                        break;
                case 0xF1 ... 0xF3:
                        if (_c_unlikely_(len < 4))
                                goto out;
                        if (_c_unlikely_(!C_UTF8_CHAR_IS_TAIL(c_load_8(str, 1))))
                                goto out;
                        if (_c_unlikely_(!C_UTF8_CHAR_IS_TAIL(c_load_8(str, 2))))
                                goto out;
                        if (_c_unlikely_(!C_UTF8_CHAR_IS_TAIL(c_load_8(str, 3))))
                                goto out;

                        str += 4;
                        len -= 4;

                        break;
                case 0xF4:
                        if (_c_unlikely_(len < 4))
                                goto out;
                        if (_c_unlikely_(c_load_8(str, 1) < 0x80 || c_load_8(str, 1) > 0x8F))
                                goto out;
                        if (_c_unlikely_(!C_UTF8_CHAR_IS_TAIL(c_load_8(str, 2))))
                                goto out;
                        if (_c_unlikely_(!C_UTF8_CHAR_IS_TAIL(c_load_8(str, 3))))
                                goto out;

                        str += 4;
                        len -= 4;

                        break;
                default:
                        goto out;
                }
        }

out:
        *strp = str;
        if (lenp)
                *lenp = len;
}
